src/main/scala/nl/vroste/zio/kinesis/client/DynamicConsumer.scala

```diff
@@ -110,13 +112,104 @@ object DynamicConsumer {
+   * and takes care of checkpointing. The other difference is that it returns a ZIO of unit rather than a ZStream.
+   *
+   * @param streamName Name of the Kinesis stream
+   * @param applicationName Application name for coordinating shard leases
+   * @param deserializer Deserializer for record values
+   * @param requestShutdown Effect that when completed will trigger a graceful shutdown of the KCL
+   *   and the streams.
+   * @param initialPosition Position in stream to start at when there is no previous checkpoint
+   *   for this application
+   * @param isEnhancedFanOut Flag for setting retrieval config - defaults to `true`. If `false` polling config is set.
+   * @param leaseTableName Optionally set the lease table name - defaults to None. If not specified the `applicationName` will be used.
+   * @param workerIdentifier Identifier used for the worker in this application group. Used in logging
+   *   and written to the lease table.
+   * @param maxShardBufferSize The maximum number of records per shard to store in a queue before blocking
+   *   the KCL record processor until records have been dequeued. Note that the stream returned from this
+   *   method will have internal chunk buffers as well.
+   * @param checkpointBatchSize Maximum number of records before checkpointing
+   * @param checkpointDuration Maximum interval before checkpointing
+   * @param recordProcessor A function for processing a `Record[T]`
+   * @tparam R ZIO environment type required by the `deserializer` and the `recordProcessor`
+   * @tparam T Type of record values
+   * @return ZIO of unit
+   */
+  def consumeWith[R, T](
+    streamName: String,
+    applicationName: String,
+    deserializer: Deserializer[R, T],
+    requestShutdown: UIO[Unit] = UIO.never,
+    initialPosition: InitialPositionInStreamExtended =
+      InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.TRIM_HORIZON),
+    isEnhancedFanOut: Boolean = true,
+    leaseTableName: Option[String] = None,
+    workerIdentifier: String = UUID.randomUUID().toString,
+    maxShardBufferSize: Int = 1024, // Prefer powers of 2
+    checkpointBatchSize: Long = 200,
+    checkpointDuration: Duration = 5.second
+  )(
+    recordProcessor: RecordProcessor[R, T]
+  ): ZIO[R with Blocking with Logging with Clock with DynamicConsumer, Throwable, Unit] =
+    for {
+      consumer <- ZIO.service[DynamicConsumer.Service]
+      _        <- consumer
+             .shardedStream(
+               streamName,
+               applicationName,
+               deserializer,
+               requestShutdown,
+               initialPosition,
+               isEnhancedFanOut,
+               leaseTableName,
+               workerIdentifier,
+               maxShardBufferSize
+             )
+             .flatMapPar(Int.MaxValue) {
+               case (_, shardStream, checkpointer) =>
+                 ZStream.fromEffect {
+                   ZIO.environment[R].flatMap { env =>
```

Comment:
I think you can change `consumeWith`s return type to be `ZIO[R with R1` where the record processor is a function of `Record[T] => RIO[R1, Unit]`. Then the `R1` environment can be defined at the place where `consumeWith` is called.

Reply:
:+1: